unidist/core/backends/mpi/core/communication.py

```diff
@@ -86,19 +102,63 @@ class MPIState:
     ----------
     comm : mpi4py.MPI.Comm
         MPI communicator.
+
+    Attributes
+    ----------
+    comm : mpi4py.MPI.Comm
+        MPI communicator.
+    host_comm : mpi4py.MPI.Comm
+        Splited MPI communicator for current host
     rank : int
         Rank of a process.
     world_sise : int
         Number of processes.
+    host : str
+        IP-address of current host
+    topology : dict
+        Dictionary, containing workers ranks assignments by IP-addresses in
+        the form: `{"node_ip0": [rank_2, rank_3, ...], "node_ip1": [rank_i, ...], ...}`.
     """
 
     __instance = None
 
-    def __init__(self, comm, rank, world_sise):
+    def __init__(self, comm):
         # attributes get actual values when MPI is initialized in `init` function
         self.comm = comm
-        self.rank = rank
-        self.world_size = world_sise
+        self.rank = comm.Get_rank()
+
+        if is_internal_host_communication_supported():
+            self.host_comm = comm.Split_type(MPI.COMM_TYPE_SHARED)
+        else:
+            self.host_comm = None
+
+        self.host = socket.gethostbyname(socket.gethostname())
+        self.world_size = comm.Get_size()
+
+        # Get topology of MPI cluster.
+        host_rank = (
+            self.host_comm.Get_rank() if self.host_comm is not None else self.rank
+        )
+        cluster_info = self.comm.allgather((self.host, self.rank, host_rank))
+
+        self.topology = defaultdict(dict)
+        self.__host_rank_by_rank = defaultdict(None)
+        self.__host_by_rank = defaultdict(None)
```

Comment:
```suggestion
        self._host_rank_by_rank = defaultdict(None)
        self._host_by_rank = defaultdict(None)
```
Add inline comments what is this for.